File: 1-starter_project_nestjs/libs/contract/src/exceptions/api.exception.ts
import { ApiErrorDto } from '../dto/api-error.dto';

export class ApiException extends Error {
  readonly statusCode: number;
  readonly apiError?: ApiErrorDto;

  constructor(statusCodeOrDto: number | ApiErrorDto, message?: string) {
    if (typeof statusCodeOrDto === 'number') {
      super(message);
      this.statusCode = statusCodeOrDto;
    } else {
      const dto = statusCodeOrDto;
      super(`API call [${dto.links}] failed, status code [${dto.httpStatusCode}], message [${dto.message}]`);
      this.statusCode = dto.httpStatusCode;
      this.apiError = dto;
    }
    this.name = 'ApiException';
  }
}

File: 1-starter_project/common/mytravels.contract/CustomException/ApiException.cs
using System.Diagnostics.CodeAnalysis;
using mytravels.contract.Dtos;

namespace mytravels.contract.CustomException;

[ExcludeFromCodeCoverage]
public class ApiException : Exception
{
    public ApiException(int statusCode, string message) : base(message)
    {
        StatusCode = statusCode;
    }

    public ApiException(ApiErrorDto apiError) : base($"API call [{apiError.Links}] failed, status code [{apiError.HttpStatusCode}], message [{apiError.Message}] ")
    {
        StatusCode = apiError.HttpStatusCode;
        ApiError = apiError;
    }

    public int StatusCode { get; }
    public ApiErrorDto ApiError { get; }

    protected ApiException(System.Runtime.Serialization.SerializationInfo serializationInfo, System.Runtime.Serialization.StreamingContext streamingContext)
    {
        throw new NotImplementedException();
    }
}